node/core/provisioner/src/disputes/prioritized_selection/mod.rs

```
@@ -0,0 +1,445 @@
+/// Please refer to `struct PartitionedDisputes` for details about the actual partitions.
+/// Each partition has got a priority implicitly assigned to it and the disputes are selected based on thus
+/// priority (e.g. disputes in partition 1, then if there is space - disputes from partition 2 and so on).
+///
+/// # Votes selection
+///
+/// Besides the prioritization described above the votes in each partition are filtered too. Provisioner
+/// fetches all onchain votes and filters them out from all partitions. As a result the Runtime receives
+/// only fresh votes (votes it didn't know about).
+///
+/// # How the onchain votes are fetched
+///
+/// The logic outlined above relies on `RuntimeApiRequest::Disputes` message from the Runtime staging API.
+/// If the staging API is not enabled - the same logic is executed with empty onchain votes set. Effectively this
+/// means that all disputes are partitioned in groups 2 or 4 and all votes are sent to the Runtime.
+pub async fn select_disputes<Sender>(
+	sender: &mut Sender,
+	metrics: &metrics::Metrics,
+	leaf: &ActivatedLeaf,
+) -> Result<MultiDisputeStatementSet, Error>
+where
+	Sender: overseer::ProvisionerSenderTrait,
+{
+	gum::trace!(
+		target: LOG_TARGET,
+		?leaf,
+		"Selecting disputes for inherent data using prioritized selection"
+	);
+
+	// Fetch the onchain disputes. We'll do a prioritization based on them.
+	let onchain = match get_onchain_disputes(sender, leaf.hash.clone()).await {
+		Ok(r) => r,
+		Err(GetOnchainDisputesError::NotSupported(runtime_api_err, relay_parent)) => {
+			// Runtime version is checked before calling this method, so the error below should never happen!
+			gum::error!(
+				target: LOG_TARGET,
+				?runtime_api_err,
+				?relay_parent,
+				"Can't fetch onchain disputes, because runtime version is not recent enough. Will continue with empty onchain disputes set.",
+			);
+			HashMap::new()
```

@eskimor maybe makes more sense to return error here instead of continuing with empty onchain votes HashMap?

---

We hope for this to not fail, but if it fails it is better to import something than nothing. In practice this should not matter as this should actually not be possible to happen at all.